framework/src/modules/dpos_v2/commands/vote_delegate.ts

```diff
@@ -184,18 +238,57 @@ export class VoteDelegateCommand extends BaseCommand {
 				// Sort account.unlocking
 				sortUnlocking(voterData.pendingUnlocks);
 
-				if (voterData.pendingUnlocks.length > MAX_UNLOCKING) {
-					throw new Error(`Pending unlocks cannot exceed ${MAX_UNLOCKING.toString()}.`);
+				if (voterData.pendingUnlocks.length > MAX_NUMBER_PENDING_UNLOCKS) {
+					this.events.get(DelegateVotedEvent).error(
+						context,
+						{
+							senderAddress,
+							delegateAddress: vote.delegateAddress,
+							amount: vote.amount,
+						},
+						PoSEventResult.VOTE_FAILED_TOO_MANY_PENDING_UNLOCKS,
+					);
+
+					throw new Error(
+						`Pending unlocks cannot exceed ${MAX_NUMBER_PENDING_UNLOCKS.toString()}.`,
+					);
+				}
+
+				await this._internalMethod.assignVoteRewards(
+					context,
+					senderAddress,
+					voterData.sentVotes[index],
+					delegateData,
+				);
+
+				// sent vote is not removed from voerData because of zero amount
+				if (aggregateVotedAmount !== BigInt(0)) {
+					voterData.sentVotes[index].amount = aggregateVotedAmount;
+					voterData.sentVotes[index].voteSharingCoefficients = delegateData.sharingCoefficients;
 				}
 			} else {
 				// Upvote amount case
-				const upvote =
-					originalUpvoteIndex > -1
-						? voterData.sentVotes[originalUpvoteIndex]
-						: {
-								delegateAddress: vote.delegateAddress,
-								amount: BigInt(0),
-						  };
+				let upvote;
+
+				if (originalUpvoteIndex > -1) {
+					upvote = voterData.sentVotes[originalUpvoteIndex];
+
+					await this._internalMethod.assignVoteRewards(
+						context.getMethodContext(),
+						senderAddress,
+						voterData.sentVotes[originalUpvoteIndex],
+						delegateData,
+					);
+
+					voterData.sentVotes[index].voteSharingCoefficients = delegateData.sharingCoefficients;
```

Comment:
This is part of the `assignVoteRewards` function according to the LIP. Not sure if that's the case in the implementation as well.